## Question 1
### What's the version of NumPy that you installed?

In [1]:
import numpy as np
np.__version__

'1.26.4'

### Getting the data

In [2]:
!wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv

--2024-05-28 18:16:30--  https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1475504 (1,4M) [text/plain]
Saving to: ‘data.csv’

data.csv            100%[===================>]   1,41M  --.-KB/s    in 0,1s    

2024-05-28 18:16:31 (10,3 MB/s) - ‘data.csv’ saved [1475504/1475504]



In [8]:
import pandas as pd
cars_df = pd.read_csv("data.csv")

## Question 2
### How many records are in the dataset?

In [7]:
cars_df.shape[0]

11914

## Question 3
### Who are the most frequent car manufacturers (top-3) according to the dataset?

In [13]:
cars_df.Make.value_counts()[:3]

Make
Chevrolet     1123
Ford           881
Volkswagen     809
Name: count, dtype: int64

## Question 4
### What's the number of unique Audi car models in the dataset?

In [27]:
cars_df.Model[cars_df.Make=="Audi"].nunique()

34

## Question 5
### How many columns in the dataset have missing values?

In [29]:
cars_df.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

## Question 6
1. Find the median value of "Engine Cylinders" column in the dataset.
2. Next, calculate the most frequent value of the same "Engine Cylinders".
3. Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
4. Now, calculate the median value of "Engine Cylinders" once again.

In [31]:
cars_df["Engine Cylinders"].median()

6.0

In [32]:
cars_df["Engine Cylinders"].value_counts()

Engine Cylinders
4.0     4752
6.0     4489
8.0     2031
12.0     230
5.0      225
10.0      68
0.0       56
3.0       30
16.0       3
Name: count, dtype: int64

In [33]:
cars_df["Engine Cylinders"].fillna(4.0, inplace=True)

In [34]:
cars_df["Engine Cylinders"].median()

6.0

### Has it changed? No

## Question 7
1. Select all the "Lotus" cars from the dataset.
2. Select only columns "Engine HP", "Engine Cylinders".
3. Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Compute the inverse of XTX.
7. Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

In [53]:
lotus_cars = cars_df[cars_df.Make=="Lotus"][["Engine HP", "Engine Cylinders"]]
X = lotus_cars.drop_duplicates().values
XTX = X.T.dot(X)
XTX_inv = np.linalg.inv(XTX)
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])
w = XTX_inv.dot(X.T).dot(y)

In [54]:
w

array([  4.59494481, -63.56432501])

### What's the value of the first element of w? 4.59494481